File: mobile/lib/domain/entities/leave_request.dart
import 'package:equatable/equatable.dart';

class LeaveRequestPreviousVersion extends Equatable {
  const LeaveRequestPreviousVersion({
    required this.leaveType,
    required this.leaveTypeLabel,
    required this.startDate,
    required this.endDate,
    required this.leaveDays,
    required this.reason,
    required this.hasSupportingDocument,
    this.supportingDocument,
    required this.isPaid,
  });

  final String leaveType;
  final String leaveTypeLabel;
  final DateTime startDate;
  final DateTime endDate;
  final int leaveDays;
  final String reason;
  final bool hasSupportingDocument;
  final String? supportingDocument;
  final bool isPaid;

  @override
  List<Object?> get props => [leaveType, startDate, endDate, leaveDays];
}

class LeaveRequestItem extends Equatable {
  const LeaveRequestItem({
    required this.id,
    required this.employeeId,
    required this.employeeName,
    required this.employeePosition,
    required this.leaveType,
    required this.leaveTypeLabel,
    required this.startDate,
    required this.endDate,
    required this.leaveDays,
    required this.reason,
    required this.status,
    required this.isPaid,
    this.policyIsPaid,
    this.isPaidOverridden = false,
    required this.hasSupportingDocument,
    this.supportingDocument,
    this.ownerRemarks,
    this.reviewedAt,
    required this.createdAt,
    this.hasPendingChanges = false,
    this.previousRequest,
  });

  final String id;
  final String employeeId;
  final String? employeeName;
  final String? employeePosition;
  final String leaveType;
  final String leaveTypeLabel;
  final DateTime startDate;
  final DateTime endDate;
  final int leaveDays;
  final String reason;
  final String status;
  final bool isPaid;
  final bool? policyIsPaid;
  final bool isPaidOverridden;
  final bool hasSupportingDocument;
  final String? supportingDocument;
  final String? ownerRemarks;
  final DateTime? reviewedAt;
  final DateTime createdAt;
  final bool hasPendingChanges;
  final LeaveRequestPreviousVersion? previousRequest;

  bool get isPending => status == 'pending';
  bool get isApproved => status == 'approved';
  bool get isRejected => status == 'rejected';
  bool get isCancellationPending => status == 'cancellation_pending';
  bool get isCancelled => status == 'cancelled';

  bool get canEdit => isPending || isApproved;
  bool get canCancel => isPending || isApproved;

  String get statusLabel => switch (status) {
        'approved' => 'Approved',
        'rejected' => 'Rejected',
        'cancellation_pending' => 'Cancellation Pending',
        'cancelled' => 'Cancelled',
        _ => 'Pending Approval',
      };

  @override
  List<Object?> get props => [id, status, leaveType, startDate, endDate];
}

const leaveTypeOptions = <(String value, String label)>[
  ('sick', 'Sick Leave'),
  ('vacation', 'Vacation Leave'),
  ('emergency', 'Emergency Leave'),
  ('maternity', 'Maternity Leave'),
  ('paternity', 'Paternity Leave'),
  ('unpaid', 'Unpaid Leave'),
  ('other', 'Other'),
];

File: backend/app/schemas/leave_request.py
from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.enums import LeaveRequestStatus, LeaveType


class LeaveRequestCreate(BaseModel):
    leave_type: LeaveType
    start_date: date
    end_date: date
    reason: str = Field(min_length=3, max_length=2000)
    supporting_document: str | None = None


class LeaveRequestUpdate(BaseModel):
    leave_type: LeaveType
    start_date: date
    end_date: date
    reason: str = Field(min_length=3, max_length=2000)
    supporting_document: str | None = None


class LeaveRequestReviewRequest(BaseModel):
    remarks: str | None = Field(default=None, max_length=500)
    # Owner override of company Leave Policy (defaults to policy when omitted).
    is_paid: bool | None = None
    override_reason: str | None = Field(default=None, max_length=500)


class LeaveRequestPreviousVersion(BaseModel):
    leave_type: LeaveType
    leave_type_label: str
    start_date: date
    end_date: date | None = None
    leave_days: int
    reason: str
    has_supporting_document: bool = False
    supporting_document: str | None = None
    is_paid: bool


class LeaveRequestResponse(BaseModel):
    id: UUID
    business_id: UUID
    employee_id: UUID
    employee_name: str | None = None
    employee_position: str | None = None
    employee_profile_image_url: str | None = None
    leave_type: LeaveType
    leave_type_label: str
    start_date: date
    end_date: date
    leave_days: int
    reason: str
    supporting_document: str | None = None
    has_supporting_document: bool = False
    status: LeaveRequestStatus
    policy_is_paid: bool
    is_paid: bool
    is_paid_overridden: bool = False
    has_pending_changes: bool = False
    previous_request: LeaveRequestPreviousVersion | None = None
    owner_remarks: str | None = None
    reviewed_by: UUID | None = None
    reviewed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}
